File: src/p2_add_two_numbers.go
/*
 * @lc app=leetcode id=2 lang=golang
 *
 * [2] Add Two Numbers
 */

package leetcode

// @lc code=start
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	head := &ListNode{}
	p := head

	for carry := 0; p != nil; p = p.Next {
		if l1 != nil {
			p.Val += l1.Val
			l1 = l1.Next
		}

		if l2 != nil {
			p.Val += l2.Val
			l2 = l2.Next
		}

		if p.Val >= 10 {
			p.Val -= 10
			carry = 1
		} else {
			carry = 0
		}

		if l1 != nil || l2 != nil || carry != 0 {
			p.Next = &ListNode{Val: carry}
		}
	}

	return head
}

// @lc code=end

File: src/p2.AddTwoNumbers.java

/*
 * @lc app=leetcode id=2 lang=java
 *
 * [2] Add Two Numbers
 */

import structs.ListNode;

class AddTwoNumbers {
    // @lc code=start
    class Solution {
        public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
            ListNode head = new ListNode(0);
            ListNode p = head;

            for (int carry = 0; p != null; p = p.next) {
                if (l1 != null) {
                    p.val += l1.val;
                    l1 = l1.next;
                }
                if (l2 != null) {
                    p.val += l2.val;
                    l2 = l2.next;
                }

                if (p.val >= 10) {
                    p.val -= 10;
                    carry = 1;
                } else {
                    carry = 0;
                }

                p.next = (l1 != null || l2 != null || carry != 0) ? new ListNode(carry) : null;
            }

            return head;
        }
    }
    // @lc code=end

    public static void main(String[] args) {
        var l1 = new ListNode(2, new ListNode(4, new ListNode(3)));
        var l2 = new ListNode(5, new ListNode(6, new ListNode(4)));
        var l3 = new ListNode(7, new ListNode(0, new ListNode(8)));

        var s = (new AddTwoNumbers()).new Solution();
        var actual = s.addTwoNumbers(l1, l2);
        System.out.println("expected: " + l3);
        System.out.println("actual: " + actual);
    }
}
